pyquil/device.py

```diff
@@ -113,158 +119,41 @@ def from_dict(d):
-
-    :ivar str isa_name: The name of the instruction set architecture for the QPU.
-    :ivar Sequence[KrausModel] gates: The tomographic estimates of all gates.
-    :ivar Dict[int,np.array] assignment_probs: The single qubit readout assignment
-        probability matrices keyed by qubit id.
-    """
-
-    def to_dict(self):
-        """
-        Create a JSON serializable representation of the noise model.
-
-        For example::
-
-            {
-                "isa_name": "example_qpu",
-                "gates": [
-                    # list of embedded dictionary representations of KrausModels here [...]
-                ]
-                "assignment_probs": {
-                    "0": [[.8, .1],
-                          [.2, .9]],
-                    "1": [[.9, .4],
-                          [.1, .6]],
-                }
-            }
-
-        :return: A dictionary representation of self.
-        :rtype: Dict[str,Any]
-        """
-        return {
-            "isa_name": self.isa_name,
-            "gates": [km.to_dict() for km in self.gates],
-            "assignment_probs": {str(qid): a.tolist() for qid, a in self.assignment_probs.items()},
-        }
-
-    @staticmethod
-    def from_dict(d):
-        """
-        Re-create the noise model from a dictionary representation.
-
-        :param Dict[str,Any] d: The dictionary representation.
-        :return: The restored noise model.
-        :rtype: NoiseModel
-        """
-        return NoiseModel(
-            isa_name=d["isa_name"],
-            gates=[KrausModel.from_dict(t) for t in d["gates"]],
-            assignment_probs={int(qid): np.array(a) for qid, a in d["assignment_probs"].items()},
-        )
-
-    def gates_by_name(self, name):
-        """
-        Return all defined noisy gates of a particular gate name.
-
-        :param str name: The gate name.
-        :return: A list of noise models representing that gate.
-        :rtype: Sequence[KrausModel]
-        """
-        return [g for g in self.gates if g.gate == name]
-
-    def __eq__(self, other):
-        return isinstance(other, NoiseModel) and self.to_dict() == other.to_dict()
-
-    def __neq__(self, other):
-        return not self.__eq__(other)
+    gates = []
+    for q in isa.qubits:
+        if q.dead:
+            continue
+        if q.type in [None, "Xhalves"]:
+            gates.extend([
+                Gate("I", [], [unpack_qubit(q.id)]),
+                Gate("RX", [np.pi / 2], [unpack_qubit(q.id)]),
+                Gate("RX", [-np.pi / 2], [unpack_qubit(q.id)]),
+                Gate("RZ", [THETA], [unpack_qubit(q.id)]),
+            ])
+        else:  # pragma no coverage
+            raise ValueError("Unknown qubit type: {}".format(q.type))
+
+    for e in isa.edges:
+        if e.dead:
+            continue
+        targets = [unpack_qubit(t) for t in e.targets]
+        if e.type in [None, "CZ", "ISWAP"]:
+            gates.append(Gate(e.type, [], targets))
```

Comment:
does this work with `None`?

Reply:
Yeah, `None` corresponds to a missing field in the ISA JSON file. cc @ecp-rigetti

Reply:
it's true that the "type" field is optional, but it defaults to "CZ" when it's not present. i think there should be a line here that does the same substitution in the event that `None` is encountered.

Reply:
Ok, I am addressing this by explicitly configuring a `DEFAULT_(QUBIT|EDGE)_TYPE` to allow adding/dropping it in de/serialization and preventing those fields to have a value of `None`.